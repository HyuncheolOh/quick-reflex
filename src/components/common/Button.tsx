import React from 'react';
import {
  TouchableOpacity,
  Text,
  StyleSheet,
  ViewStyle,
  TextStyle,
  ActivityIndicator,
} from 'react-native';
import { COLORS, SPACING, TYPOGRAPHY, BORDER_RADIUS, SHADOWS } from '../../constants';

interface ButtonProps {
  title: string;
  onPress: () => void;
  variant?: 'primary' | 'secondary' | 'danger' | 'ghost';
  size?: 'small' | 'medium' | 'large';
  disabled?: boolean;
  loading?: boolean;
  style?: ViewStyle;
  textStyle?: TextStyle;
}

export const Button: React.FC<ButtonProps> = ({
  title,
  onPress,
  variant = 'primary',
  size = 'medium',
  disabled = false,
  loading = false,
  style,
  textStyle,
}) => {
  const buttonStyle = [
    styles.base,
    styles[`${variant}Container`],
    styles[`${size}Container`],
    disabled && styles.disabledContainer,
    style,
  ];

  const titleStyle = [
    styles.baseText,
    styles[`${variant}Text`],
    styles[`${size}Text`],
    disabled && styles.disabledText,
    textStyle,
  ];

  return (
    <TouchableOpacity
      style={buttonStyle}
      onPress={onPress}
      disabled={disabled || loading}
      activeOpacity={0.8}
    >
      {loading ? (
        <ActivityIndicator
          size="small"
          color={variant === 'ghost' ? COLORS.PRIMARY : COLORS.TEXT_PRIMARY}
        />
      ) : (
        <Text style={titleStyle}>{title}</Text>
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  base: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: BORDER_RADIUS.LG,
    ...SHADOWS.SMALL,
  },
  
  // Container variants
  primaryContainer: {
    backgroundColor: COLORS.PRIMARY,
  },
  secondaryContainer: {
    backgroundColor: COLORS.SECONDARY,
  },
  dangerContainer: {
    backgroundColor: COLORS.ERROR,
  },
  ghostContainer: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: COLORS.PRIMARY,
    ...SHADOWS.SMALL,
  },
  
  // Container sizes
  smallContainer: {
    paddingHorizontal: SPACING.MD,
    paddingVertical: SPACING.SM,
    minHeight: 36,
  },
  mediumContainer: {
    paddingHorizontal: SPACING.LG,
    paddingVertical: SPACING.MD,
    minHeight: 44,
  },
  largeContainer: {
    paddingHorizontal: SPACING.XL,
    paddingVertical: SPACING.LG,
    minHeight: 52,
  },
  
  // Disabled state
  disabledContainer: {
    opacity: 0.5,
  },
  
  // Base text
  baseText: {
    textAlign: 'center',
    fontWeight: TYPOGRAPHY.FONT_WEIGHT.SEMIBOLD,
  },
  
  // Text variants
  primaryText: {
    color: COLORS.TEXT_PRIMARY,
  },
  secondaryText: {
    color: COLORS.TEXT_PRIMARY,
  },
  dangerText: {
    color: COLORS.TEXT_PRIMARY,
  },
  ghostText: {
    color: COLORS.PRIMARY,
  },
  
  // Text sizes
  smallText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.SM,
  },
  mediumText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.MD,
  },
  largeText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.LG,
  },
  
  // Disabled text
  disabledText: {
    opacity: 0.7,
  },
});