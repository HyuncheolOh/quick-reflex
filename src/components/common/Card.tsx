import React from 'react';
import {
  View,
  StyleSheet,
  ViewStyle,
  TouchableOpacity,
  TouchableOpacityProps,
} from 'react-native';
import { COLORS, SPACING, BORDER_RADIUS, SHADOWS } from '../../constants';

interface CardProps extends TouchableOpacityProps {
  children: React.ReactNode;
  style?: ViewStyle;
  padding?: keyof typeof SPACING;
  touchable?: boolean;
  variant?: 'default' | 'elevated' | 'outlined';
}

export const Card: React.FC<CardProps> = ({
  children,
  style,
  padding = 'MD',
  touchable = false,
  variant = 'default',
  ...touchableProps
}) => {
  const cardStyle = [
    styles.base,
    styles[variant],
    { padding: SPACING[padding] },
    style,
  ];

  if (touchable) {
    return (
      <TouchableOpacity style={cardStyle} activeOpacity={0.8} {...touchableProps}>
        {children}
      </TouchableOpacity>
    );
  }

  return <View style={cardStyle}>{children}</View>;
};

const styles = StyleSheet.create({
  base: {
    borderRadius: BORDER_RADIUS.LG,
  },
  
  default: {
    backgroundColor: COLORS.CARD,
  },
  
  elevated: {
    backgroundColor: COLORS.CARD,
    ...SHADOWS.MEDIUM,
  },
  
  outlined: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: COLORS.TEXT_TERTIARY,
  },
});