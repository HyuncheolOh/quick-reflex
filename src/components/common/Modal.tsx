import React from 'react';
import {
  Modal as RNModal,
  View,
  StyleSheet,
  Dimensions,
  TouchableOpacity,
  Text,
} from 'react-native';
import { COLORS, SPACING, TYPOGRAPHY, BORDER_RADIUS } from '../../constants';
import { Button } from './Button';

const { width, height } = Dimensions.get('window');

interface ModalProps {
  visible: boolean;
  onClose: () => void;
  title?: string;
  children: React.ReactNode;
  showCloseButton?: boolean;
  animationType?: 'slide' | 'fade' | 'none';
}

export const Modal: React.FC<ModalProps> = ({
  visible,
  onClose,
  title,
  children,
  showCloseButton = true,
  animationType = 'fade',
}) => {
  return (
    <RNModal
      visible={visible}
      transparent
      animationType={animationType}
      onRequestClose={onClose}
    >
      <View style={styles.overlay}>
        <TouchableOpacity
          style={styles.backdrop}
          activeOpacity={1}
          onPress={onClose}
        />
        
        <View style={styles.container}>
          {title && (
            <View style={styles.header}>
              <Text style={styles.title}>{title}</Text>
              {showCloseButton && (
                <TouchableOpacity onPress={onClose} style={styles.closeButton}>
                  <Text style={styles.closeButtonText}>✕</Text>
                </TouchableOpacity>
              )}
            </View>
          )}
          
          <View style={styles.content}>
            {children}
          </View>
        </View>
      </View>
    </RNModal>
  );
};

export const ConfirmModal: React.FC<{
  visible: boolean;
  onClose: () => void;
  onConfirm: () => void;
  title: string;
  message: string;
  confirmText?: string;
  cancelText?: string;
  variant?: 'danger' | 'primary';
}> = ({
  visible,
  onClose,
  onConfirm,
  title,
  message,
  confirmText = '확인',
  cancelText = '취소',
  variant = 'primary',
}) => {
  return (
    <Modal visible={visible} onClose={onClose} title={title} showCloseButton={false}>
      <Text style={styles.message}>{message}</Text>
      
      <View style={styles.buttonContainer}>
        <Button
          title={cancelText}
          onPress={onClose}
          variant="ghost"
          style={styles.button}
        />
        <Button
          title={confirmText}
          onPress={() => {
            onConfirm();
            onClose();
          }}
          variant={variant === 'danger' ? 'danger' : 'primary'}
          style={styles.button}
        />
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  overlay: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  
  backdrop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  
  container: {
    backgroundColor: COLORS.SURFACE,
    borderRadius: BORDER_RADIUS.XL,
    maxWidth: width - SPACING.XL * 2,
    maxHeight: height - SPACING.XXL * 2,
    minWidth: 280,
  },
  
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: SPACING.LG,
    paddingTop: SPACING.LG,
    paddingBottom: SPACING.MD,
    borderBottomWidth: 1,
    borderBottomColor: COLORS.TEXT_TERTIARY,
  },
  
  title: {
    fontSize: TYPOGRAPHY.FONT_SIZE.LG,
    fontWeight: TYPOGRAPHY.FONT_WEIGHT.BOLD,
    color: COLORS.TEXT_PRIMARY,
    flex: 1,
  },
  
  closeButton: {
    padding: SPACING.SM,
    marginLeft: SPACING.MD,
  },
  
  closeButtonText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.LG,
    color: COLORS.TEXT_SECONDARY,
  },
  
  content: {
    padding: SPACING.LG,
  },
  
  message: {
    fontSize: TYPOGRAPHY.FONT_SIZE.MD,
    color: COLORS.TEXT_PRIMARY,
    textAlign: 'center',
    marginBottom: SPACING.LG,
    lineHeight: 22,
  },
  
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    gap: SPACING.MD,
  },
  
  button: {
    flex: 1,
  },
});