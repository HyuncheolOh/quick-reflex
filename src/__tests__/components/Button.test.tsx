import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Button } from '../../components/common/Button';

// Mock useThemedColors hook
jest.mock('../../hooks', () => ({
  useThemedColors: () => ({
    PRIMARY: '#007AFF',
    SECONDARY: '#34C759',
    ERROR: '#FF3B30',
    TEXT_PRIMARY: '#FFFFFF',
  }),
}));

describe('Button', () => {
  const defaultProps = {
    title: 'Test Button',
    onPress: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Basic Rendering', () => {
    it('should render button with title', () => {
      const { getByText } = render(<Button {...defaultProps} />);
      
      expect(getByText('Test Button')).toBeTruthy();
    });

    it('should call onPress when tapped', () => {
      const mockOnPress = jest.fn();
      const { getByText } = render(
        <Button {...defaultProps} onPress={mockOnPress} />
      );
      
      fireEvent.press(getByText('Test Button'));
      expect(mockOnPress).toHaveBeenCalledTimes(1);
    });
  });

  describe('Variants', () => {
    it('should render primary variant by default', () => {
      const { getByText } = render(<Button {...defaultProps} />);
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: '#007AFF',
          }),
        ])
      );
    });

    it('should render secondary variant', () => {
      const { getByText } = render(
        <Button {...defaultProps} variant="secondary" />
      );
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: '#34C759',
          }),
        ])
      );
    });

    it('should render danger variant', () => {
      const { getByText } = render(
        <Button {...defaultProps} variant="danger" />
      );
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: '#FF3B30',
          }),
        ])
      );
    });

    it('should render ghost variant', () => {
      const { getByText } = render(
        <Button {...defaultProps} variant="ghost" />
      );
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: 'transparent',
            borderWidth: 1,
            borderColor: '#007AFF',
          }),
        ])
      );
    });
  });

  describe('Sizes', () => {
    it('should render medium size by default', () => {
      const { getByText } = render(<Button {...defaultProps} />);
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            minHeight: 44,
          }),
        ])
      );
    });

    it('should render small size', () => {
      const { getByText } = render(<Button {...defaultProps} size="small" />);
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            minHeight: 36,
          }),
        ])
      );
    });

    it('should render large size', () => {
      const { getByText } = render(<Button {...defaultProps} size="large" />);
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            minHeight: 52,
          }),
        ])
      );
    });
  });

  describe('States', () => {
    it('should be disabled when disabled prop is true', () => {
      const mockOnPress = jest.fn();
      const { getByText } = render(
        <Button {...defaultProps} onPress={mockOnPress} disabled />
      );
      const button = getByText('Test Button').parent;
      
      expect(button?.props.disabled).toBe(true);
      
      fireEvent.press(getByText('Test Button'));
      expect(mockOnPress).not.toHaveBeenCalled();
    });

    it('should show loading indicator when loading', () => {
      const { getByTestId, queryByText } = render(
        <Button {...defaultProps} loading />
      );
      
      expect(queryByText('Test Button')).toBeNull();
      // ActivityIndicator should be present but testing it requires additional setup
    });

    it('should be disabled when loading', () => {
      const mockOnPress = jest.fn();
      const { getByText } = render(
        <Button {...defaultProps} onPress={mockOnPress} loading />
      );
      
      // Since loading shows ActivityIndicator instead of text, we need to find the touchable differently
      const touchable = getByText('Test Button').parent?.parent; // TouchableOpacity is parent of the content
      expect(touchable?.props.disabled).toBe(true);
    });

    it('should apply disabled opacity', () => {
      const { getByText } = render(<Button {...defaultProps} disabled />);
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            opacity: 0.6,
          }),
        ])
      );
    });
  });

  describe('Custom Styles', () => {
    it('should apply custom container style', () => {
      const customStyle = { marginTop: 20 };
      const { getByText } = render(
        <Button {...defaultProps} style={customStyle} />
      );
      const button = getByText('Test Button').parent;
      
      expect(button?.props.style).toEqual(
        expect.arrayContaining([customStyle])
      );
    });

    it('should apply custom text style', () => {
      const customTextStyle = { fontSize: 20 };
      const { getByText } = render(
        <Button {...defaultProps} textStyle={customTextStyle} />
      );
      const text = getByText('Test Button');
      
      expect(text.props.style).toEqual(
        expect.arrayContaining([customTextStyle])
      );
    });
  });


  describe('Snapshot Tests', () => {
    it('should match snapshot for primary button', () => {
      const { toJSON } = render(<Button {...defaultProps} />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot for ghost button', () => {
      const { toJSON } = render(<Button {...defaultProps} variant="ghost" />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot for disabled button', () => {
      const { toJSON } = render(<Button {...defaultProps} disabled />);
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot for loading button', () => {
      const { toJSON } = render(<Button {...defaultProps} loading />);
      expect(toJSON()).toMatchSnapshot();
    });
  });
});