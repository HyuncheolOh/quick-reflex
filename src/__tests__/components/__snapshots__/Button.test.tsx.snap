// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Snapshot Tests should match snapshot for disabled button 1`] = `
<TouchableOpacity
  activeOpacity={0.8}
  disabled={true}
  onPress={[MockFunction]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderRadius": 12,
        "elevation": 3,
        "justifyContent": "center",
        "minHeight": 44,
        "paddingHorizontal": 24,
        "paddingVertical": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.22,
        "shadowRadius": 2.22,
      },
      {
        "opacity": 0.6,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "opacity": 0.6,
        },
        undefined,
      ]
    }
  >
    Test Button
  </Text>
</TouchableOpacity>
`;

exports[`Button Snapshot Tests should match snapshot for ghost button 1`] = `
<TouchableOpacity
  activeOpacity={0.8}
  disabled={false}
  onPress={[MockFunction]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#007AFF",
        "borderRadius": 12,
        "borderWidth": 1,
        "elevation": 3,
        "justifyContent": "center",
        "minHeight": 44,
        "paddingHorizontal": 24,
        "paddingVertical": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.22,
        "shadowRadius": 2.22,
      },
      false,
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#007AFF",
          "fontSize": 16,
          "fontWeight": "600",
          "textAlign": "center",
        },
        false,
        undefined,
      ]
    }
  >
    Test Button
  </Text>
</TouchableOpacity>
`;

exports[`Button Snapshot Tests should match snapshot for primary button 1`] = `
<TouchableOpacity
  activeOpacity={0.8}
  disabled={false}
  onPress={[MockFunction]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#007AFF",
        "borderRadius": 12,
        "elevation": 3,
        "justifyContent": "center",
        "minHeight": 44,
        "paddingHorizontal": 24,
        "paddingVertical": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.22,
        "shadowRadius": 2.22,
      },
      false,
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#FFFFFF",
          "fontSize": 16,
          "fontWeight": "600",
          "textAlign": "center",
        },
        false,
        undefined,
      ]
    }
  >
    Test Button
  </Text>
</TouchableOpacity>
`;
